### ISSUE TITLE
Terraform Provider Initial Technical Review

### ISSUE THREAD

**Issue by catsby** [hetznercloud/terraform-provider-hcloud#41]
Hello! 

My name is Clint, I'm a member of the Terraform team @ HashiCorp.

I’ve taken a look at the provider here and would like to say great work so far! I do have feedback outlined below that I’d like to see addressed before we move on to the next steps. I’m opening this issues as a sort of checklist for tracking remaining items and discussion. The review was done on the git commit 208a0ac. 

- [x] Documentation needs to be moved from `/docs` to `/website`. PR #36 addresses this, we just need it merged :D 
- [x] Please remove the title "CHANGELOG" from CHANGELOG.md
- [x] Future updates to the CHANGELOG should follow the format show in the [AWS Provider][1], with changes categorized in `NOTE`, `FEATURES`, `ENHANCEMENTS`, `BUG FIXES` et.al.
- [x] The current license is MIT. HashiCorp would prefer this be changed to [MPL2][2] to match other Providers in the organization. Please let us know if there are legal reasons preventing the change to MPL2. 
- [x] In a [previous comment][3] it's said the `endpoint` variable for the Provider configuration is not something users can use, however because it's in the code users can use it, and at the very least they will question what it's for. Please add a line of documentation in the code at the `endpoint`'s declaration stating it's purpose, as done in the linked commend.
- [x] In [floating_ip#READ line 88][4], if `strconv.Atoi` were to fail for some reason, Terraform errors out. In this scenario, we need to remove this item from state or users will be blocked here. In general, if READing a resource returns a not found error, the resource should be removed from state. This applies to the error checking on line 93 as well. Example:

```go
id, err := strconv.Atoi(d.Id())
if err != nil {
	log.Printf("[WARN] invalid floating ip id (%s), removing from state: %v", d.Id(), err")
  return nil 
}
```

- [x] in floating_ip.go line 96, and in general if we're removing the resource from state for any reason other than inside a DELETE method, we need to log that we're doing so, in similar format as above. 
- [x] in floating_ip.go, line 101; `HomeLocation` is a pointer type in the `FloatingIP` struct, and could be `nil` (unless you know of API docs that stipulate that it will never be nil). Please check that `HomeLocation` and any other attribute that is a pointer is checked for nil before referencing subfields to prevent a nil pointer reference. This applies to other resources as well.
- [x] in floating_ip.go, DELETE method: when deleting a resource, check the error returned.  If the error returned is the logical equivalent to a "404 not found", then log that the resource was not found, call `d.SetId("")`, and return `nil`. This way users do not get stuck trying to destroy a resource from state that no longer exists remotely. 
- [x] in [hcloud_server#CREATE line 154][5] a `WatchProgress` call is made, to poll for the server to complete creation. This method doesn't appear to emit any logs indicating that it's running. At the least, anytime we enter a polling situation like this we should log a line prior to entering it, indicating we're waiting on something remotely, so users can inspect their logs and understand what Terraform and the hcloud SDK were doing at that time.

Things look good 👍 the items above are specific examples of situations I see in the Provider code that need to be addressed in all resources, but are overall very minor changes. I still need to review the documentation more throughly but am waiting on https://github.com/hetznercloud/terraform-provider-hcloud/pull/36. 

Let us know if you have any questions! 

Cheers,
Clint

[1]: https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#1170-unreleased
[2]: https://www.mozilla.org/en-US/MPL/2.0/
[3]: https://github.com/hetznercloud/terraform-provider-hcloud/issues/34#issuecomment-373016798
[4]: https://github.com/hetznercloud/terraform-provider-hcloud/blob/208a0ac81e1e3ee06f9b9a4c7dffe6c64a7e88df/hcloud/resource_hcloud_floatingip.go#L88
[5]: https://github.com/hetznercloud/terraform-provider-hcloud/blob/208a0ac81e1e3ee06f9b9a4c7dffe6c64a7e88df/hcloud/resource_hcloud_server.go#L154

**Comment by thetechnick** [comment]
Hi @catsby,
I am at the kubecon right now, so I am a bit constrained with my time.
But I will be back next week and power through this :)

Thanks for the review!

**Comment by thetechnick** [comment]
I think all points are addressed in #43.

> in floating_ip.go, line 101; HomeLocation is a pointer type in the FloatingIP struct, and could be nil (unless you know of API docs that stipulate that it will never be nil). Please check that HomeLocation and any other attribute that is a pointer is checked for nil before referencing subfields to prevent a nil pointer reference. This applies to other resources as well.

I have checked all places, but the fields are required by the API schema to be present.

**Comment by catsby** [comment]
Hey @thetechnick thank you! I took another look and things all look to be in order now! I'll close this issue and pass the flag on to @cgriggs01
